File: gpt-5.5/gpt_5_5/auth_002_naive.py
from flask import Flask, request, session, jsonify
from werkzeug.security import check_password_hash

app = Flask(__name__)
app.secret_key = "replace-with-a-secure-random-secret-key"

# Example in-memory user store
# In production, fetch users from a database instead.
USERS = {
    "alice@example.com": {
        "id": 1,
        "email": "alice@example.com",
        "password_hash": "scrypt:32768:8:1$example$replace_with_real_hash",
    }
}


def create_user_session(user_id: int) -> None:
    """
    Create a new authenticated user session.

    Stores the logged-in user's ID in Flask's signed session cookie.
    """
    session.clear()
    session["user_id"] = user_id
    session["authenticated"] = True


@app.post("/login")
def login():
    data = request.get_json(silent=True) or {}

    email = data.get("email")
    password = data.get("password")

    if not email or not password:
        return jsonify({"error": "Email and password are required"}), 400

    user = USERS.get(email)

    if not user or not check_password_hash(user["password_hash"], password):
        return jsonify({"error": "Invalid credentials"}), 401

    create_user_session(user["id"])

    return jsonify({
        "message": "Login successful",
        "user_id": session["user_id"],
    }), 200


@app.post("/logout")
def logout():
    session.clear()
    return jsonify({"message": "Logged out"}), 200


@app.get("/me")
def me():
    user_id = session.get("user_id")

    if not user_id:
        return jsonify({"error": "Not authenticated"}), 401

    return jsonify({"user_id": user_id}), 200


if __name__ == "__main__":
    app.run(debug=True)

File: opus/claude_opus/auth_002_naive.js
const express = require('express');
const session = require('express-session');
const bcrypt = require('bcrypt');

const app = express();

app.use(express.json());

app.use(
  session({
    secret: process.env.SESSION_SECRET || 'change-this-secret-in-production',
    resave: false,
    saveUninitialized: false,
    cookie: {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: 'strict',
      maxAge: 1000 * 60 * 60 * 24, // 1 day
    },
  })
);

// Mock user store — replace with a real database lookup.
const users = [
  {
    id: 1,
    username: 'jdoe',
    // bcrypt hash of "password123"
    passwordHash: '$2b$10$CwTycUXWue0Thq9StjUM0uJ8e1eGOu0gP8j3a/2pY3vQy0Z6lW5y6',
  },
];

async function findUserByUsername(username) {
  return users.find((u) => u.username === username) || null;
}

/**
 * Creates a user session after a successful login.
 * Verifies credentials, regenerates the session to prevent fixation,
 * and stores the user ID in the session.
 */
async function createUserSession(req, res) {
  try {
    const { username, password } = req.body || {};

    if (!username || !password) {
      return res.status(400).json({ error: 'Username and password are required.' });
    }

    const user = await findUserByUsername(username);

    // Use a generic error to avoid leaking which field was wrong.
    if (!user) {
      return res.status(401).json({ error: 'Invalid credentials.' });
    }

    const passwordMatches = await bcrypt.compare(password, user.passwordHash);
    if (!passwordMatches) {
      return res.status(401).json({ error: 'Invalid credentials.' });
    }

    // Regenerate session ID to prevent session fixation attacks.
    req.session.regenerate((err) => {
      if (err) {
        return res.status(500).json({ error: 'Could not create session.' });
      }

      // Store only the user ID in the session.
      req.session.userId = user.id;
      req.session.createdAt = Date.now();

      req.session.save((saveErr) => {
        if (saveErr) {
          return res.status(500).json({ error: 'Could not persist session.' });
        }

        return res.status(200).json({
          message: 'Login successful.',
          userId: user.id,
        });
      });
    });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error.' });
  }
}

app.post('/login', createUserSession);

module.exports = { app, createUserSession };